File: workstation/src/main/java/io/xj/model/meme/MemeConstellation.java
// Copyright (c) 1999-2022, XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.model.meme;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 Compute normalized string representation of an unordered set of memes
 for the purpose of identifying unique constellations.
 <p>
 for each unique sequence-pattern-meme constellation within the main sequence https://github.com/xjmusic/workstation/issues/208
 */
public enum MemeConstellation {
  ;
  static final String CONSTELLATION_DELIMITER = "_";

  /**
   @return unique constellation for this set of names
   */
  public static String fromNames(Collection<String> names) {
    Map<String, Boolean> uniqueNames = new HashMap<>();
    names.forEach(meme -> uniqueNames.put(meme, true));
    Set<String> pieces = new TreeSet<>(uniqueNames.keySet());
    return String.join(CONSTELLATION_DELIMITER, pieces);
  }

  /**
   @return set of names parsed from constellation
   */
  public static Collection<String> toNames(String constellation) {
    return List.of(constellation.split(CONSTELLATION_DELIMITER));
  }
}

File: engine/src/meme/MemeConstellation.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/meme/MemeConstellation.h"

using namespace XJ;

std::string MemeConstellation::fromNames(const std::set<std::string> &names) {
  std::set<std::string> uniqueNames;
  for (const auto &meme: names) {
    uniqueNames.emplace(meme);
  }
  return join(CONSTELLATION_DELIMITER, uniqueNames);
}

std::set<std::string> MemeConstellation::toNames(const std::string &constellation)  {
  return split(constellation, CONSTELLATION_DELIMITER);
}

std::string MemeConstellation::join(const std::string &delimiter, const std::set<std::string> &pieces) {
  std::ostringstream joined;
  std::copy(pieces.begin(), pieces.end(), std::ostream_iterator<std::string>(joined, delimiter.c_str()));
  const std::string result = joined.str();
  return result.substr(0, result.size() - delimiter.size()); // remove the last delimiter
}

std::set<std::string> MemeConstellation::split(const std::string &str, const std::string &delimiter)  {
  std::set<std::string> tokens;
  size_t start = 0, end;
  while ((end = str.find(delimiter, start)) != std::string::npos) {
    tokens.insert(str.substr(start, end - start));
    start = end + delimiter.size();
  }
  tokens.insert(str.substr(start));
  return tokens;
}

const std::string MemeConstellation::CONSTELLATION_DELIMITER = "_";
